sanic/websocket.py

```diff
@@ -7,23 +10,748 @@
+        if self.close_connection_task is not None and not self.close_connection_task.done():
+            if self.data_finished_fut is not None and not self.data_finished_fut.done():
+                self.data_finished_fut.cancel()
+            # Don't close, auto_close_connection will take care of it.
+            return False
+        SanicProtocol.close(self.io_proto)
+        return True
+
+    async def auto_close_connection(self) -> None:
+        """
+        Close the WebSocket Connection
+        When the opening handshake succeeds, :meth:`connection_open` starts
+        this coroutine in a task. It waits for the data transfer phase to
+        complete then it closes the TCP connection cleanly.
+        When the opening handshake fails, :meth:`fail_connection` does the
+        same. There's no data transfer phase in that case.
+        """
+        try:
+            # Wait for the data transfer phase to complete.
+            if self.data_finished_fut is not None:
+                try:
+                    await self.data_finished_fut
+                except asyncio.CancelledError:
+                    pass
+
+            # Cancel the keepalive ping task.
+            if self.keepalive_ping_task is not None:
+                self.keepalive_ping_task.cancel()
+
+            # Half-close the TCP connection if possible (when there's no TLS).
+            if self.io_proto.transport is not None and self.io_proto.transport.can_write_eof():
+                error_logger.warning("Websocket half-closing TCP connection")
+                self.io_proto.transport.write_eof()
+                if self.connection_lost_waiter is not None:
+                    if await self.wait_for_connection_lost(timeout=0):
+                        return
+        finally:
+            # The try/finally ensures that the transport never remains open,
+            # even if this coroutine is cancelled (for example).
+            if self.connection_lost_waiter is not None and self.connection_lost_waiter.done():
+                if self.io_proto.transport is None or self.io_proto.transport.is_closing():
+                    return
+            SanicProtocol.close(self.io_proto)
+            if self.connection_lost_waiter is not None:
+                await self.wait_for_connection_lost()
+                if self.connection_lost_waiter.done():
+                    return
+                error_logger.warning("Timeout waiting for TCP connection to close. Aborting")
+                if self.io_proto.transport is not None:
+                    self.io_proto.transport.abort()
+
+    def abort_pings(self) -> None:
+        """
+        Raise ConnectionClosed in pending keepalive pings.
+        They'll never receive a pong once the connection is closed.
+        """
+        assert self.connection.state is CLOSED
+
+        for ping in self.pings.values():
+            ping.set_exception(ConnectionClosedError(1006, ""))
```

Comment:
After https://github.com/aaugustin/websockets/pull/998 you should change this to `ConnectionClosedError(None, None)`. It's a backwards-incompatible change affecting users who create `ConnectionClosed` exceptions, which doesn't happen often.

Reply:
@aaugustin On a separate note... is this a recent thing that `ConnectionClosedError` subclasses `Exception`? I feel like it had been `CancelledError`.

Reply:
Well, my memory could fail me 😎 but as far as I remember the websockets exception hierarchy as always rooted at `class WebSocketException(Exception): ...`.

Reply:
@aaugustin 
Just clarifying here. Do you mean use `ConnectionClosedError(1006, None)`?
The first argument to ConnectionClosedError must be an int, it cannot be None.